In [1]:
import pandas as pd
from pathlib import Path

# Loading in the data

## Betting Data

In [13]:
betting_data_2018 = Path('./data/2018.xlsx')
original_betting_data_2018_df = pd.read_excel(betting_data_2018)
original_betting_data_2018_df.head()

ATP  Location              Tournament       Date  Series    Court Surface  \
0    1  Brisbane  Brisbane International 2017-12-31  ATP250  Outdoor    Hard   
1    1  Brisbane  Brisbane International 2017-12-31  ATP250  Outdoor    Hard   
2    1  Brisbane  Brisbane International 2018-01-01  ATP250  Outdoor    Hard   
3    1  Brisbane  Brisbane International 2018-01-01  ATP250  Outdoor    Hard   
4    1  Brisbane  Brisbane International 2018-01-01  ATP250  Outdoor    Hard   

       Round  Best of         Winner  ...    EXW   EXL   LBW   LBL   PSW  \
0  1st Round        3  Dolgopolov O.  ...   2.10  1.65  2.20  1.67  2.24   
1  1st Round        3   De Minaur A.  ...   2.80  1.38  2.75  1.44  3.14   
2  1st Round        3    Harrison R.  ...   1.62  2.15  1.67  2.20  1.66   
3  1st Round        3       Ebden M.  ...   2.35  1.53  2.37  1.57  2.38   
4  1st Round        3      Zverev M.  ...   1.47  2.50  1.57  2.37  1.44   

    PSL  MaxW  MaxL  AvgW  AvgL  
0  1.71  2.29  1.77  2.17  1.68  
1  1.41  3.25  1.44  2.96  1.39  
2  2.35  1.75  2.35  1.67  2.17  
3  1.64  2.52  1.65  2.37  1.58  
4  3.00  1.57  3.00  1.44  2.73  

[5 rows x 40 columns]

## Match Data

In [14]:
match_data_2018 = Path('./data/atp_matches_2018.csv')
original_match_data_2018_df = pd.read_csv(match_data_2018)
original_match_data_2018_df.head()

tourney_id tourney_name surface  draw_size tourney_level  tourney_date  \
0  2018-M020     Brisbane    Hard         32             A      20180101   
1  2018-M020     Brisbane    Hard         32             A      20180101   
2  2018-M020     Brisbane    Hard         32             A      20180101   
3  2018-M020     Brisbane    Hard         32             A      20180101   
4  2018-M020     Brisbane    Hard         32             A      20180101   

   match_num  winner_id  winner_seed winner_entry    ...    w_bpFaced l_ace  \
0        300     106401          3.0          NaN    ...          5.0   8.0   
1        299     106401          3.0          NaN    ...          2.0   5.0   
2        298     105992          NaN          NaN    ...          6.0   9.0   
3        297     105777          1.0          NaN    ...          2.0   6.0   
4        296     106401          3.0          NaN    ...          3.0   6.0   

   l_df l_svpt  l_1stIn  l_1stWon  l_2ndWon  l_SvGms  l_bpSaved l_bpFaced  
0   3.0   58.0     32.0      22.0      12.0      9.0        4.0       7.0  
1   6.0   74.0     44.0      32.0      16.0     13.0        3.0       6.0  
2   5.0   94.0     62.0      49.0      16.0     16.0        0.0       3.0  
3   1.0   89.0     53.0      39.0      23.0     15.0        4.0       6.0  
4   3.0   75.0     41.0      33.0      16.0     13.0        2.0       4.0  

[5 rows x 49 columns]

# Exploring and Understanding the Data

Now that we have the data loaded, we have to figure out what information this data gives us (i.e. what the columns represent).

## Betting data

Let's look at the [betting data documentation](http://www.tennis-data.co.uk/notes.txt) (shown below) to see what we can gather. The data includes betting odds for the matches by different betting entities, but we will ignore those for now and just focus on the match data.

<img src="./data/data_legend.png">

To get a better understanding of the data, let's take a look at the numerical summary pandas gives us:

In [15]:
original_betting_data_2018_df.describe()

ATP      Best of        WRank        LRank          WPts  \
count  2637.000000  2637.000000  2637.000000  2636.000000   2637.000000   
mean     33.760334     3.384528    60.269245    87.683991   1732.301100   
std      18.554584     0.788308    73.188987   113.492047   1815.483268   
min       1.000000     3.000000     1.000000     1.000000      1.000000   
25%      19.000000     3.000000    19.000000    35.000000    694.000000   
50%      33.000000     3.000000    46.000000    64.000000   1051.000000   
75%      50.000000     3.000000    80.000000   102.000000   1905.000000   
max      67.000000     5.000000  1821.000000  1821.000000  10600.000000   

               LPts           W1           L1           W2           L2  \
count   2636.000000  2622.000000  2623.000000  2607.000000  2607.000000   
mean    1126.762519     5.802441     4.208921     5.764480     4.030687   
std     1062.844898     1.261901     1.799389     1.279958     1.834431   
min        1.000000     0.000000     0.000000     0.000000     0.000000   
25%      559.750000     6.000000     3.000000     6.000000     3.000000   
50%      835.500000     6.000000     4.000000     6.000000     4.000000   
75%     1285.000000     6.000000     6.000000     6.000000     6.000000   
max    10600.000000     7.000000     7.000000     7.000000     7.000000   

          ...               EXW         EXL          LBW          LBL  \
count     ...       2618.000000  2618.00000  2162.000000  2162.000000   
mean      ...          1.827762     2.96500     1.855148     3.045023   
std       ...          0.895751     2.13104     0.873949     2.529572   
min       ...          1.010000     1.02000     1.000000     1.040000   
25%       ...          1.300000     1.69250     1.330000     1.730000   
50%       ...          1.570000     2.30000     1.610000     2.250000   
75%       ...          2.100000     3.35000     2.100000     3.400000   
max       ...         17.000000    20.00000    11.000000    41.000000   

               PSW          PSL         MaxW         MaxL         AvgW  \
count  2617.000000  2617.000000  2636.000000  2636.000000  2636.000000   
mean      1.926335     3.463091     1.991089     3.677830     1.863961   
std       1.065434     3.414277     1.150385     3.925328     0.942274   
min       1.010000     1.030000     1.020000     1.050000     1.010000   
25%       1.340000     1.770000     1.370000     1.810000     1.320000   
50%       1.620000     2.460000     1.670000     2.550000     1.590000   
75%       2.190000     3.600000     2.260000     3.720000     2.120000   
max      18.030000    33.900000    20.000000    52.250000    14.070000   

              AvgL  
count  2636.000000  
mean      3.183194  
std       2.713412  
min       1.030000  
25%       1.720000  
50%       2.360000  
75%       3.390000  
max      25.410000  

[8 rows x 30 columns]

This table tells us a few things:
- There are 67 tournaments represented in this dataset (the max of `ATP` is 67)
- There may be missing values in `LRank` and in the `W1` through `L5` columns (the `count` of these columns is less than the maximum value of `count` for any column)


## Match data

There is no official documentation for the match data, but the columns titles are fairly self-explanatory. Let's look at the pandas numerical summary for the match data.

In [16]:
original_match_data_2018_df.describe()

draw_size  tourney_date    match_num      winner_id  winner_seed  \
count  2494.000000  2.494000e+03  2494.000000    2494.000000   948.000000   
mean     61.589415  2.018048e+07   231.905774  109729.885726     8.126582   
std      44.142876  2.485647e+02    91.467684   14857.070038     7.624144   
min       4.000000  2.018010e+07     1.000000  100644.000000     1.000000   
25%      32.000000  2.018022e+07   200.000000  104898.000000     3.000000   
50%      32.000000  2.018043e+07   275.000000  105526.000000     6.000000   
75%     128.000000  2.018072e+07   289.000000  106298.000000    10.000000   
max     128.000000  2.018092e+07   701.000000  208029.000000    32.000000   

         winner_ht   winner_age  winner_rank  winner_rank_points  \
count  1660.000000  2494.000000  2487.000000         2487.000000   
mean    186.912048    27.628326    88.418979         1603.599920   
std       7.986847     4.630301   163.833733         1793.900975   
min     163.000000    14.926762     1.000000            1.000000   
25%     183.000000    23.986995    21.000000          638.000000   
50%     185.000000    27.994524    51.000000          981.000000   
75%     190.000000    31.102669    91.000000         1757.000000   
max     208.000000    39.378508  1821.000000        10600.000000   

            loser_id     ...         w_bpFaced        l_ace         l_df  \
count    2494.000000     ...       2473.000000  2473.000000  2473.000000   
mean   111484.771451     ...          5.129802     5.594015     3.572988   
std     18561.077638     ...          3.992374     5.435946     2.606244   
min    100644.000000     ...          0.000000     0.000000     0.000000   
25%    104925.000000     ...          2.000000     2.000000     2.000000   
50%    105583.000000     ...          4.000000     4.000000     3.000000   
75%    106378.000000     ...          7.000000     8.000000     5.000000   
max    207973.000000     ...         25.000000    61.000000    18.000000   

            l_svpt      l_1stIn     l_1stWon     l_2ndWon      l_SvGms  \
count  2473.000000  2473.000000  2473.000000  2473.000000  2473.000000   
mean     84.093409    50.841892    34.318237    15.439143    11.650222   
std      30.258697    19.844166    15.126202     7.247338     5.223811   
min      11.000000     6.000000     1.000000     1.000000     0.000000   
25%      61.000000    36.000000    24.000000    10.000000     9.000000   
50%      79.000000    48.000000    32.000000    14.000000    11.000000   
75%     101.000000    62.000000    43.000000    20.000000    15.000000   
max     291.000000   218.000000   171.000000    50.000000    50.000000   

         l_bpSaved    l_bpFaced  
count  2473.000000  2473.000000  
mean      4.863728     8.677719  
std       3.332023     4.233006  
min       0.000000     0.000000  
25%       2.000000     6.000000  
50%       4.000000     8.000000  
75%       7.000000    11.000000  
max      22.000000    28.000000  

[8 rows x 35 columns]

### Missing data

From the previous table, we can see there are a lot of missing values in the data (the `count` of some columns is less than the maximum value of `count` for all columns). Let's examine these records (matches).

In [17]:
original_match_data_2018_df[original_match_data_2018_df.isnull().any(axis=1)] # display all rows that have missing data in any column

tourney_id                 tourney_name surface  \
0                           2018-M020                     Brisbane    Hard   
1                           2018-M020                     Brisbane    Hard   
2                           2018-M020                     Brisbane    Hard   
3                           2018-M020                     Brisbane    Hard   
4                           2018-M020                     Brisbane    Hard   
5                           2018-M020                     Brisbane    Hard   
6                           2018-M020                     Brisbane    Hard   
7                           2018-M020                     Brisbane    Hard   
8                           2018-M020                     Brisbane    Hard   
9                           2018-M020                     Brisbane    Hard   
10                          2018-M020                     Brisbane    Hard   
11                          2018-M020                     Brisbane    Hard   
12                          2018-M020                     Brisbane    Hard   
13                          2018-M020                     Brisbane    Hard   
14                          2018-M020                     Brisbane    Hard   
15                          2018-M020                     Brisbane    Hard   
16                          2018-M020                     Brisbane    Hard   
17                          2018-M020                     Brisbane    Hard   
18                          2018-M020                     Brisbane    Hard   
19                          2018-M020                     Brisbane    Hard   
20                          2018-M020                     Brisbane    Hard   
21                          2018-M020                     Brisbane    Hard   
22                          2018-M020                     Brisbane    Hard   
23                          2018-M020                     Brisbane    Hard   
24                          2018-M020                     Brisbane    Hard   
25                          2018-M020                     Brisbane    Hard   
26                          2018-M020                     Brisbane    Hard   
27                          2018-0451                         Doha    Hard   
28                          2018-0451                         Doha    Hard   
29                          2018-0451                         Doha    Hard   
...                               ...                          ...     ...   
2464   2018-M-DC-2018-WG-M-USA-BEL-01  Davis Cup WG QF: USA vs BEL    Hard   
2465   2018-M-DC-2018-WG-M-USA-SRB-01  Davis Cup WG R1: USA vs SRB    Clay   
2466   2018-M-DC-2018-WG-M-USA-SRB-01  Davis Cup WG R1: USA vs SRB    Clay   
2467   2018-M-DC-2018-WG-M-USA-SRB-01  Davis Cup WG R1: USA vs SRB    Clay   
2468  2018-M-DC-2018-WG-PO-ARG-COL-01  Davis Cup WG PO: ARG vs COL    Clay   
2469  2018-M-DC-2018-WG-PO-ARG-COL-01  Davis Cup WG PO: ARG vs COL    Clay   
2470  2018-M-DC-2018-WG-PO-ARG-COL-01  Davis Cup WG PO: ARG vs COL    Clay   
2471  2018-M-DC-2018-WG-PO-AUT-AUS-01  Davis Cup WG PO: AUT vs AUS    Clay   
2472  2018-M-DC-2018-WG-PO-AUT-AUS-01  Davis Cup WG PO: AUT vs AUS    Clay   
2473  2018-M-DC-2018-WG-PO-AUT-AUS-01  Davis Cup WG PO: AUT vs AUS    Clay   
2474  2018-M-DC-2018-WG-PO-CAN-NED-01  Davis Cup WG PO: CAN vs NED    Hard   
2475  2018-M-DC-2018-WG-PO-CAN-NED-01  Davis Cup WG PO: CAN vs NED    Hard   
2476  2018-M-DC-2018-WG-PO-CAN-NED-01  Davis Cup WG PO: CAN vs NED    Hard   
2477  2018-M-DC-2018-WG-PO-GBR-UZB-01  Davis Cup WG PO: GBR vs UZB    Hard   
2478  2018-M-DC-2018-WG-PO-GBR-UZB-01  Davis Cup WG PO: GBR vs UZB    Hard   
2479  2018-M-DC-2018-WG-PO-GBR-UZB-01  Davis Cup WG PO: GBR vs UZB    Hard   
2480  2018-M-DC-2018-WG-PO-HUN-CZE-01  Davis Cup WG PO: HUN vs CZE    Clay   
2481  2018-M-DC-2018-WG-PO-HUN-CZE-01  Davis Cup WG PO: HUN vs CZE    Clay   
2482  2018-M-DC-2018-WG-PO-HUN-CZE-01  Davis Cup WG PO: HUN vs CZE    Clay   
2483  2018-M-DC-2018-WG-PO-HUN-CZE-01  Davis Cup 

It seems like the `winner_seed` and `winner_entry` columns have a lot of missing data, but there may be more elsewhere. Let's see how many missing entries each column has.

In [18]:
original_match_data_2018_df.apply(lambda col: col.isnull().sum(), axis=0)

tourney_id               0
tourney_name             0
surface                  0
draw_size                0
tourney_level            0
tourney_date             0
match_num                0
winner_id                0
winner_seed           1546
winner_entry          2171
winner_name              0
winner_hand              4
winner_ht              834
winner_ioc               0
winner_age               0
winner_rank              7
winner_rank_points       7
loser_id                 0
loser_seed            1909
loser_entry           2000
loser_name               0
loser_hand              11
loser_ht              1015
loser_ioc                0
loser_age                3
loser_rank              31
loser_rank_points       31
score                    0
best_of                  0
round                    0
minutes                 12
w_ace                   21
w_df                    21
w_svpt                  21
w_1stIn                 21
w_1stWon                21
w_2ndWon                21
w

As expected, the `seed` and `entry` columns for both winner and losers have a lot of missing values. Fortunately, the `entry` column isn't really useful information anyway, and the `winner_rank` and `loser_rank` columns give us a way to consider seeding in our model. Therefore, we can safely drop the `seed` and `entry` columns. Additionally, the `ht` (height) of the winner and the loser have many missing values; since the height of a player isn't a strong feature in predicting the outcome of a match, we can also safely drop the `ht` column.

In [8]:
drop_cols = ['winner_seed', 'loser_seed', 'winner_entry', 'loser_entry', 'winner_ht', 'loser_ht']
match_data_2018_df = original_match_data_2018_df.drop(drop_cols, axis=1)
match_data_2018_df.head()

tourney_id tourney_name surface  draw_size tourney_level  tourney_date  \
0  2018-M020     Brisbane    Hard         32             A      20180101   
1  2018-M020     Brisbane    Hard         32             A      20180101   
2  2018-M020     Brisbane    Hard         32             A      20180101   
3  2018-M020     Brisbane    Hard         32             A      20180101   
4  2018-M020     Brisbane    Hard         32             A      20180101   

   match_num  winner_id      winner_name winner_hand    ...     w_bpFaced  \
0        300     106401     Nick Kyrgios           R    ...           5.0   
1        299     106401     Nick Kyrgios           R    ...           2.0   
2        298     105992    Ryan Harrison           R    ...           6.0   
3        297     105777  Grigor Dimitrov           R    ...           2.0   
4        296     106401     Nick Kyrgios           R    ...           3.0   

   l_ace  l_df  l_svpt  l_1stIn l_1stWon l_2ndWon l_SvGms  l_bpSaved  \
0    8.0   3.0    58.0     32.0     22.0     12.0     9.0        4.0   
1    5.0   6.0    74.0     44.0     32.0     16.0    13.0        3.0   
2    9.0   5.0    94.0     62.0     49.0     16.0    16.0        0.0   
3    6.0   1.0    89.0     53.0     39.0     23.0    15.0        4.0   
4    6.0   3.0    75.0     41.0     33.0     16.0    13.0        2.0   

   l_bpFaced  
0        7.0  
1        6.0  
2        3.0  
3        6.0  
4        4.0  

[5 rows x 43 columns]

It's also worth noting that the match stats (e.g. `w_ace`, `l_bpSaved`, etc.) all have 21 missing values, which are likely the same 21 matches. Let's confirm that this is the case.

In [9]:
match_stats = ['w_ace', 'w_df', 'w_svpt', 'w_1stIn', 'w_1stWon', 'w_2ndWon', 'w_SvGms', 'w_bpSaved', 'w_bpFaced',
              'l_ace', 'l_df', 'l_svpt', 'l_1stIn', 'l_1stWon', 'l_2ndWon', 'l_SvGms', 'l_bpSaved', 'l_bpFaced']

In [10]:
# display the match stats for the 21 matches that have missing values for `w_ace`
missing_match_stats = match_data_2018_df[match_data_2018_df['w_ace'].isnull()][match_stats]
missing_match_stats

w_ace  w_df  w_svpt  w_1stIn  w_1stWon  w_2ndWon  w_SvGms  w_bpSaved  \
27      NaN   NaN     NaN      NaN       NaN       NaN      NaN        NaN   
28      NaN   NaN     NaN      NaN       NaN       NaN      NaN        NaN   
112     NaN   NaN     NaN      NaN       NaN       NaN      NaN        NaN   
406     NaN   NaN     NaN      NaN       NaN       NaN      NaN        NaN   
622     NaN   NaN     NaN      NaN       NaN       NaN      NaN        NaN   
872     NaN   NaN     NaN      NaN       NaN       NaN      NaN        NaN   
1453    NaN   NaN     NaN      NaN       NaN       NaN      NaN        NaN   
1573    NaN   NaN     NaN      NaN       NaN       NaN      NaN        NaN   
1799    NaN   NaN     NaN      NaN       NaN       NaN      NaN        NaN   
1881    NaN   NaN     NaN      NaN       NaN       NaN      NaN        NaN   
2050    NaN   NaN     NaN      NaN       NaN       NaN      NaN        NaN   
2274    NaN   NaN     NaN      NaN       NaN       NaN      NaN        NaN   
2321    NaN   NaN     NaN      NaN       NaN       NaN      NaN        NaN   
2322    NaN   NaN     NaN      NaN       NaN       NaN      NaN        NaN   
2353    NaN   NaN     NaN      NaN       NaN       NaN      NaN        NaN   
2389    NaN   NaN     NaN      NaN       NaN       NaN      NaN        NaN   
2393    NaN   NaN     NaN      NaN       NaN       NaN      NaN        NaN   
2412    NaN   NaN     NaN      NaN       NaN       NaN      NaN        NaN   
2413    NaN   NaN     NaN      NaN       NaN       NaN      NaN        NaN   
2417    NaN   NaN     NaN      NaN       NaN       NaN      NaN        NaN   
2420    NaN   NaN     NaN      NaN       NaN       NaN      NaN        NaN   

      w_bpFaced  l_ace  l_df  l_svpt  l_1stIn  l_1stWon  l_2ndWon  l_SvGms  \
27          NaN    NaN   NaN     NaN      NaN       NaN       NaN      NaN   
28          NaN    NaN   NaN     NaN      NaN       NaN       NaN      NaN   
112         NaN    NaN   NaN     NaN      NaN       NaN       NaN      NaN   
406         NaN    NaN   NaN     NaN      NaN       NaN       NaN      NaN   
622         NaN    NaN   NaN     NaN      NaN       NaN       NaN      NaN   
872         NaN    NaN   NaN     NaN      NaN       NaN       NaN      NaN   
1453        NaN    NaN   NaN     NaN      NaN       NaN       NaN      NaN   
1573        NaN    NaN   NaN     NaN      NaN       NaN       NaN      NaN   
1799        NaN    NaN   NaN     NaN      NaN       NaN       NaN      NaN   
1881        NaN    NaN   NaN     NaN      NaN       NaN       NaN      NaN   
2050        NaN    NaN   NaN     NaN      NaN       NaN       NaN      NaN   
2274        NaN    NaN   NaN     NaN      NaN       NaN       NaN      NaN   
2321        NaN    NaN   NaN     NaN      NaN       NaN       NaN      NaN   
2322        NaN    NaN   NaN     NaN      NaN       NaN       NaN      NaN   
2353        NaN    NaN   NaN     NaN      NaN       NaN       NaN      NaN   
2389        NaN    NaN   NaN     NaN      NaN       NaN       NaN      NaN   
2393        NaN    NaN   NaN     NaN      NaN       NaN       NaN      NaN   
2412        NaN    NaN   NaN     NaN      NaN       NaN       NaN      NaN   
2413        NaN    NaN   NaN     NaN      NaN       NaN       NaN      NaN   
2417        NaN    NaN   NaN     NaN      NaN       NaN       NaN      NaN   
2420        NaN    NaN   NaN     NaN      NaN       NaN       NaN      NaN   

      l_bpSaved  l_bpFaced  
27          NaN        NaN  
28          NaN        NaN  
112         NaN        NaN  
406         NaN        NaN  
622         NaN        NaN  
872         NaN        NaN  
1453        NaN        NaN  
1573        NaN        NaN  
1799        NaN        NaN  
1881        NaN        NaN  
2050        NaN        NaN  
2274        NaN        NaN  
2321        NaN        NaN  
2322        NaN        NaN  
2353        NaN        NaN  
2389        NaN        NaN  
2393        NaN        NaN  
2412        NaN        NaN  
2413        NaN      

Since all the values in this table are `NaN`, we can conclude that these 21 matches are the only matches that have missing values for the match stats. Match stats will likely be important features in our model, so it is worth dropping matches that do not have data for these features. Moreover, since only 21 out of 2637 matches are missing these data, we can safely drop these 21 matches without worrying about causing a significant harm to the quality of our training data.

In [11]:
match_data_2018_df= match_data_2018_df.drop(missing_match_stats.index, axis=0)